### ISSUE TITLE
Could not read base quality scores

### ISSUE THREAD

**Issue by elcortegano** [google/deepvariant#434]
**Describe the issue:**
The prints that read base quality scores cannot be read, as result, no variants are reported. However, I can visualize these values in the reads in IGV. How is that these values cannot be read? This is the line with the error, which repeats one after.

    2021-03-26 19:12:43.550815: W third_party/nucleus/io/sam_reader.cc:534] Could not read base quality scores m64036_210113_122249/147655225/ccs: Not found: Could not read base quality scores

**Setup**
  - Operative system: Ubuntu 20.04
  - DeepVariant version: 1.1.0 (latest)
  - Installation method: docker
  - Type of data: PacBio HiFi. BAM files aligned to the reference with `minimap2 -ax map-pb`.

**Steps to reproduce:**
  - Command:
```
  docker run \
    -v /home/user/working_directory:/input \
    -v /home/user/working_directory:/output \
    google/deepvariant:1.1.0 \
    /opt/deepvariant/bin/run_deepvariant \
    --model_type=PACBIO \
    --ref=/input/reference.fa \
    --reads=/input/file.bam \
    --output_vcf=/output/file.vcf \
    --call_variants_extra_args="use_openvino=true" \
    --num_shards=4 \
    --logging_dir=/output/logs
```

**Does the quick start test work on your system?**
Yes. The test works without problem.

**Comment by AndrewCarroll** [comment]
Hi @elcortegano 

There are a few potential causes for this. First, you may want to extract this particular read and compare the number of bases in it against the number of base quality values (these will need to match).

samtools view ${BAM} | grep m64036_210113_122249\/147655225\/ccs

should retrieve this particular read.

One other thing which can cause this error is if the BAM file is truncated (for example if a download ran out of space). Could you check both of these (that the file isn't truncated, and that the read doesn't have an obvious problem) and provide a bit more information?

Thank you,
Andrew

**Comment by elcortegano** [comment]
The file does not look truncated (according to `md5sum` values).

The reads seem to have as many bases as quality values. Once of the reads reported in my data is "m64036_210113_122249/1898/ccs", the result of samtools view + grep on this read is the attached (
[read.txt](https://github.com/google/deepvariant/files/6223181/read.txt)).

Reads come from PacBio HiFi, and were aligned to the reference from the reads FASTQ files with:

```
minimap2 -t 32 -ax map-pb <reference.fa> <reads.fq> <file.sam> && \
samtools view -@ 32 -S -o <tmp.bam> <file.sam> && \
samtools calmd -@ 32 <tmp.bam> <reference.fa> <file.md.bam> && \
samtools sort -@ 32 -o <file.bam> <file.md.bam> && \
samtools index -@ 32 <file.bam>
```

**Comment by AndrewCarroll** [comment]
Hi @elcortegano 

Thank you for that analysis. Are you able to share the local snippet for this BAM file by running samtools view -bh for the coordinate in question and share it here? I would like to reproduce the issue to see what is going on.

Thank you,
Andrew

**Comment by elcortegano** [comment]
Hi Andrew,

I am not sure how to provide the BAM file you request, if I run:

    samtools view -bh <file.bam> <read>  >  file_read.bam 

An error appears indicating that the read specifies an invalid region or unknown reference. Could you please provide a more detailed command?

Thank you
Eugenio

**Comment by AndrewCarroll** [comment]
Hi @elcortegano 

I was suggesting to query the region of the BAM where that read is present. You can extract the read itself with:

samtools view ${BAM} | grep m64036_210113_122249\/147655225\/ccs

In the string that comes out from that, the 3rd tab-separated field is the chromosome which the read is mapped to. The 4th tab-separated field is the position of the genome that the read maps at. I was suggesting to take the chromosome and position, and extract only reads overlapping that position with:

samtools view -bh ${BAM} ${CHR}:${POS} > subset.bam.

Having either the read itself, or the subset would be useful. If you would like to share this with an email address you can send either to awcarroll@google.com 

Thank you,
Andrew

**Comment by elcortegano** [comment]
Got it, thank you Andrew,

A zip file with the read info and the subset BAMs is around 300 MB in size. Is there any other way I can send you these files?

Thank you,
Eugenio

**Comment by AndrewCarroll** [comment]
Hi @elcortegano 

Interesting, that is a lot larger than I expected. But I ran a test and got a similar result with a PacBio BAM. 

Can you attach just that single read here as a BAM file (using the grep command). I can likely run a test from that. If that fails, the best way might be to upload to a Google Drive object and share permissions with me for the 300MB file.

Thank you for your patience,
Andrew

**Comment by elcortegano** [comment]
Attached a file for a different read named `m64036_210113_122249/1898/ccs`. The plant text file contains the  result of:

    samtools view file.bam | grep m64036_210113_122249/1898/ccs > read.txt

If I am understanding this right, there are 4 genomic regions containing that read aligned. One of these is located at chromosome_08:4578858. The BAM file attached corresponds to that coordinate (it resulted in the smallest BAM file), generated with:

    samtools view -bh file.bam chromosome_08:4578858 > chr8_pos4578858.bam

Attachment is: [chr8.zip](https://github.com/google/deepvariant/files/6273294/chr8.zip)

**Comment by AndrewCarroll** [comment]
Hi @elcortegano 

Thank you for the file, it's very informative, and I have a few observations that might help.

First, all of the reads in this file which have sequence bases do have the correct number of quality values. The reads are not malformatted.

However, there are reads with the SAM flag 256, secondary alignment. This occurs when the mapper finds a place which is almost or as good to map the reads to. Not all mappers report secondary alignments, and this is often controlled by a parameter. These reads have neither sequence bases, nor quality values, but do have a CIGAR string. I believe (but am not certain) that DeepVariant is attempting to parse these reads, and this is causing the error. I was not able to confirm, since I wasn't able to find the reference genome used to map. These reads can be validly ignored by DeepVariant.

If this is the case, there are a few options to proceed. First, I believe that if you perform the command: 

```
samtools view -bh -f 256 file.bam > new_file.bam
```

The file should now work with DeepVariant. 

One thing it might be good to consider, based on my inspection of the CIGAR strings, I think these are likely HiFi reads. The program tag for minimap indicates the parameter -ax map-pb. I think that parameter is optimized for CLR. I believe the parameter for CCS/HiFi is -ax asm20.

We usually take mapped BAM files from pbmm2, which wraps minimap2 with parameters optimized for HiFi. This could be why we didn't notice this exact issue before. If this is the source of your problem, our team can fix this behavior in future releases by ignoring flag 256 reads. You may want to consider mapping with pbmm2 as well (https://github.com/PacificBiosciences/pbmm2)

If this does not fix your issue, could you please point me to the reference genome which you used to map to, I would need to try running DeepVariant and walk through the error more closely.

Thank you,
Andrew

**Comment by elcortegano** [comment]
Hi Andrew, thank you very much for the feedback. This is something new I have learnt about the BAM files.

Using the filtered BAM file, the error message disappears. The number of variants called has also increased considerably (~x20 for variants with PASS tag).

Our reads are in fact HiFi. We have been doing the alignment with `minimap2 -ax map-pb` because to our understanding `deepvariant` is designed for read alignments (and not assembly-to-reference alignments as achieved with `minimap2 -ax asm`). Is this a misunderstanding? Could `deepvariant` be safely used with BAMs for assembly-to-reference alignments?

Thank you again,
Eugenio

**Comment by williamrowell** [comment]
Hello Eugenio,

- I agree that it looks like the secondary alignments are causing problems, since neither the actual read sequence nor the base qualities are stored in the BAM records for secondary alignments.  When the parser hits these rows, it seems to be having trouble parsing the `*` base quality string as individual values.
- The `minimap2` parameters `-x map-pb` and `-x asm` refer to alignment parameters, and don't make any assumptions on what types of reads are being aligned (e.g. reads-to-reference vs assembly-to-reference).
- The PacBio model for DeepVariant has been trained on reads-to-reference alignments with pbmm2.
- I would highly recommend aligning your HiFi reads for DeepVariant with [pbmm2](https://github.com/PacificBiosciences/pbmm2).  In addition to the alignment presets (which I discuss in the next point), we have some post-alignment filters that are applied.  It's also just easier to use with PacBio data.  To align a human HiFi uBAM (`hifi_reads.bam`) to a reference for downstream variant calling, I use:  `pbmm2 align --num-threads 24 --preset HIFI --sort -c 0 -y 70 --sample <sample_name> reference.fasta hifi_reads.bam aligned.bam`.  This produces an aligned, sorted BAM, with all of the fields and tags necessary to be processed by DeepVariant.  If you don't have a local SMRTLink installation, you can install pbmm2 with `conda install -c bioconda pbmm2`
- The `pbmm2 --preset HIFI` parameters are _roughly_ equivalent to these minimap parameters: `-x map-pb -a --eqx -L -O 5,56 -E 4,1 -B 5 --secondary=no -z 400,50 -r 2k -Y`.  Notice that we don't use homopolymer compressed minimizers (`-H`).

Billy

**Comment by AndrewCarroll** [comment]
Hi @elcortegano 

Billy has provided some excellent, detailed answers. I'll just add a little bit of context to your question 

"Our understanding ```deepvariant``` is designed for read alignments (and not assembly-to-reference alignments as achieved with ```minimap2 -ax asm```"

This is correct, DeepVariant is for read alignments assembly-to-reference alignment. Because minimap2 has original written when most PacBio data was CLR, the choice of the word ```map-pb``` was chosen for this present. When CCS became more common, a new recommendation was made by minimap2 authors to use the parameters for assembly with an sequence divergence tuned to HiFi data (this is the 20 part of ```--asm20```). If you search for the string ``ccs``` in the minimap2 github, this should pull up the line.

So in short, DeepVariant is for  read alignments, and the parameter for read mapping HiFi reads happens to manifest in minimap2 as ```minimap2 -ax asm20```.

**Comment by elcortegano** [comment]
I see! I am very grateful for the support. Not only the problem is solved, it is everything much more clear now, and I have learned a lot from your feedback. Thank you.

**Comment by AndrewCarroll** [comment]
Hi @elcortegano 

I'm glad you were able to resolve the issue and had a positive experience. Thank you for your patience in the process, and for providing the data needed to debug it. I'll close the issue for now, but please feel free to ask about anything else if the need arises.
